README.md
[![CRAN](http://www.r-pkg.org/badges/version/ODataQuery)](http://cran.r-project.org/package=ODataQuery)
[![status](https://tinyverse.netlify.com/badge/ODataQuery)](https://CRAN.R-project.org/package=ODataQuery)
[![Downloads](http://cranlogs.r-pkg.org/badges/ODataQuery)](http://cran.r-project.org/package=ODataQuery)


# ODataQuery #
Query on R OData

This package aims to make OData services more accessible to R users.

## Installation ##

Cran version:

```R
install.packages("ODataQuery")
```


Development version:

```R
install.packages("devtools")
devtools::install_github("https://github.com/lverweijen/ODataQuery")
```

## Example usage ##

```R
# Initialisation
url <- "https://services.odata.org/V4/TripPinServiceRW"
trip_service <- ODataQuery$new(url)
people_entity <- trip_service$path("People")

# Find all people whose name starts with an R
people_entity$
  select("UserName", "FirstName", "LastName")$
  filter(to_odata(startsWith(FirstName, "R")))$
  all()

# Find a person named Scott
people_entity$
  filter(to_odata(FirstName == "Scott"))$
  one()

# Find Scott's friends
people_entity$
  get("scottketchum")$
  path("Friends")$
  all()
```

See vignettes [demo](vignettes/demo.Rmd) and [querying](vignettes/querying.Rmd) for more examples.

## Other R packages dealing with OData ##

- [OData](https://CRAN.R-project.org/package=OData) - very basic
- [cbsodataR](https://CRAN.R-project.org/package=cbsodataR) - for data from Statistics, the Netherlands in particular
- [odataR](https://github.com/HanOostdijk/odataR/)



NEWS.md
# ODataQuery 0.5.3

* Concessions to get it published on cran:
- By default print, doesn't print any records.
- `prefix_functions` accepts dummy dot-arguments but raises an error when called with more than one arguments

# ODataQuery 0.5.2

* Make all fields of ODataQuery private except `url`.
* Constructor arguments `.resource` and `.query_options` that should not be used directly get a leading dot.
* ODataQuery is no longer cloneable (It's immutable now anyway).

# ODataQuery 0.5.1

* `top` and `skip` get 10 as default parameter.
* `print` by default prints the whole data.frame.

# ODataQuery 0.5.0

* New function `to_odata` that translates R code for use in an OData query.
* `and_query`, `or_query` and `not_query` now return an `odata_query`.

# ODataQuery 0.4.0

* Change interaction with jsonlite.
Functions that retrieve data no longer take additional arguments that are passed to jsonlite.
Additional parameters to jsonlite can from now on be supplied by a parameter `jsonlite_args`.

* Pass additional options to `httr::GET`.
Additional parameters to jsonlite can from now on be supplied by a parameter `httr_args`.
This can be useful if you are behind a proxyserver or if authentication is required.

* Method `retrieve` on ODataQuery takes a new parameter `count`.


R/nse_tests.R
# nse_tests

# to_odata(5 + 4)

# to_odata(3 <= x)

# to_odata(5 + 4 * 3 - 2 / 7 == (a - b) / c * d %% 23)
# "5 add 4 mul 3 sub 2 divby 7 eq (a sub b) divby c mul d mod 23"

# to_odata(a == 3 and b != 6 || d < 9 && not lala >= bar)
# [1] "a eq 3 and b ne 6 or d lt 9 and not(lala ge bar)"

# to_odata(paste(a, "b", c, sep=", "))
# "concat(concat(concat(concat(a,', '),'b'),', '),c)"

# to_odata(all(Friends, f ~ f$FirstName == 'John'))
# "Friends.all(f:f.FirstName eq 'John')"

# to_odata(any(Friends, f ~ f$FirstName == 'John'))
# "Friends.any(f:f.FirstName eq 'John')"

# to_odata(a$b$c - (3))
# "a.b.c sub (3)"

# > to_odata(startsWith(Naam, "R_"))
# [1] "startswith(Naam,'R_')"

# to_odata(a(b(c(x, 2), 3), 4))
# "a(b(c(x,2),3),4)"

# > to_odata(list(1, a, 2, 3, "hello"))
# [1] "[1,a,2,3,\"hello\"]"

# to_odata(a + b + c)

# process_dollar(rlang::expr(a$b$c$d$e(3$f$g)))
# a$b$c$d$e(`3/f/g`)

# Name in ('Milk', 'Cheese')
# > to_odata(Name %in% list('Milk', 'Cheese'))
# odata_expr:  Name in ["Milk","Cheese"]

# > to_odata(5 + "chicken")
# odata_expr:  5 add 'chicken'
# > to_odata(5 %add% "chicken")
# odata_expr:  5 add 'chicken'

# > to_odata(list(FirstName, LastName) %in% list(list("John","Doe"), list("Jane","Smith")))
# odata_expr:  [FirstName,LastName] in [["John","Doe"],["Jane","Smith"]]

# Error
# > to_odata(Items$any(d ~ d$Quantity > 100))
# odata_expr:  Items/any(d:d/Quantity gt 100)

# Good
# to_odata(any(d ~ d$Quantity > 100))
# odata_expr:  any(d:d.Quantity gt 100)

# > to_odata(5:3)
# odata_expr:  [5,4,3]

# > to_odata("hello" %in% c("hello", "goodbye"))
# odata_expr:  'hello' in ("hello","goodbye")

# > to_odata(c(hello) == c("hello"))
# odata_expr:  (hello) eq ("hello")

tests/tinytest.R

if ( requireNamespace("tinytest", quietly=TRUE) ){
  tinytest::test_package("ODataQuery")
}



inst/tinytest/test_querying.R
library(ODataQuery)

# Function to check
expect_equal_character <- function(x, y) {
  expect_equal(as.character(x), as.character(y))
}

## Test literals with to_odata_
expect_equal_character(
  to_odata_(5),
  "5")

expect_equal_character(
  to_odata_("5"),
  "'5'")

expect_equal_character(
  to_odata_(TRUE),
  "true")

expect_equal_character(
  to_odata_(FALSE),
  "false")

expect_equal_character(
  to_odata_(NULL),
  "null")

expect_equal_character(
  to_odata_(list(1, "lala")),
  "[1,\"lala\"]")

expect_equal_character(
  to_odata_(c(1)),
  "1")

## Test literals with to_odata
expect_equal_character(
  to_odata(c(1)),
  "(1)")

expect_equal_character(
  to_odata(c(1, 2, "hello")),
  "(1,2,'hello')")

expect_equal_character(
  to_odata(1:5),
  "[1,2,3,4,5]")

expect_equal_character(
  to_odata(5:3),
  "[5,4,3]")

## Test binary operators
expect_equal_character(
  to_odata(5 + 4),
  "5 add 4")

expect_equal_character(
  to_odata(3 <= x),
  "3 le x"
)

expect_equal_character(
  to_odata(a + b + d),
  "a add b add d")

expect_equal_character(
  to_odata(5 + 4 * 3 - 2 / 7 == (a - b) / e * d %% 23),
  "5 add 4 mul 3 sub 2 divby 7 eq (a sub b) divby e mul d mod 23")

expect_equal_character(
  to_odata((1 + 2) * (3 + 4)),
  "(1 add 2) mul (3 add 4)")

expect_equal_character(
  to_odata(a == 3 && b != 6 || d < 9 && ! lala >= bar),
  "a eq 3 and b ne 6 or d lt 9 and not(lala ge bar)")

## Test strings
expect_equal_character(
  to_odata(paste(a, "b", e, sep=", ")),
  "concat(concat(concat(concat(a,', '),'b'),', '),e)")

expect_equal_character(
  to_odata(startsWith(Naam, "R_")),
  "startswith(Naam,'R_')")

## Test attributes
expect_equal_character(
  to_odata(Friends$all(f ~ f$FirstName == 'John')),
  "Friends/all(f:f/FirstName eq 'John')")

expect_equal_character(
  to_odata(Friends$any(f ~ f$FirstName == 'John')),
  "Friends/any(f:f/FirstName eq 'John')")

expect_equal_character(
  to_odata(a$b$d - (3)),
  "a/b/d sub (3)")

expect_equal_character(
  to_odata(a(b(d(x, 2), 3), 4)),
  "a(b(d(x,2),3),4)")

expect_equal_character(
  to_odata(a$b$h$d$e(3$f$g)),
  "a/b/h/d/e(3/f/g)")

expect_equal_character(
  to_odata(Items$any(d ~ d$Quantity > 100)),
  "Items/any(d:d/Quantity gt 100)")

## Test lists
expect_equal_character(
  to_odata(list(1, a, 2, 3, "hello")),
  "[1,a,2,3,\"hello\"]")

expect_equal_character(
  to_odata(Name %in% list('Milk', 'Cheese')),
  "Name in [\"Milk\",\"Cheese\"]")

expect_equal_character(
  to_odata(list(FirstName, LastName) %in% list(list("John","Doe"), list("Jane","Smith"))),
  "[FirstName,LastName] in [[\"John\",\"Doe\"],[\"Jane\",\"Smith\"]]")

## Other tests
expect_equal_character(
  to_odata(5 + "chicken"),
  "5 add 'chicken'")

expect_equal_character(
  to_odata(5 %add% "chicken"),
  "5 add 'chicken'")

expect_equal_character(
  to_odata("hello" %in% c("hello", "goodbye")),
  "'hello' in ('hello','goodbye')")

expect_equal_character(
  to_odata(c(hello) == c("hello")),
  "(hello) eq ('hello')")

## Quick queries
expect_equal_character(
  and_query("Column eq OtherColumn",
            FirstName.startswith = 'A',
            LastName.eq = 'Scott'),
  "(Column eq OtherColumn and startswith(FirstName,'A') and LastName eq 'Scott')")

expect_equal_character(
  or_query("ExpireDate eq null",
           ExpireDate.lt = "2020-07-07"),
  "(ExpireDate eq null or ExpireDate lt '2020-07-07')")

expect_equal_character(
 not_query(or_query(Age.lt = 21, Age.gt = 65)),
 "not ((Age lt 21 or Age gt 65))")


R/retrieve.R
#' Retrieve data
#'
#' @param url Which url to fetch data from
#' @param metadata Whether and how metadata is included
#' @param httr_args List of additional arguments passed on to httr::GET
#' @param jsonlite_args List of additional arguments passed on to jsonlite::fromJSON
#' @return Data including metadata
#' @export
#' @family retrieve
#'
#' @examples
#' \dontrun{
#' url <- "https://services.odata.org/V4/TripPinServiceRW"
#' retrieve_data(url)
#' }
retrieve_data <- function(url,
                          metadata = c("none", "minimal", "all"),
                          httr_args = list(),
                          jsonlite_args = list()) {
  metadata <- match.arg(metadata)
  accept <- paste0("application/json;odata.metadata=", metadata)
  useragent <- "https://github.com/lverweijen/odata_r"

  args <- list(url, httr::add_headers(Accept = accept, UserAgent = useragent))
  req <- do.call(httr::GET, c(args, httr_args))
  httr::stop_for_status(req)

  json <- httr::content(req, as = "text", encoding = "UTF-8")
  parsed <- do.call(jsonlite::fromJSON, c(json, jsonlite_args))
  parsed
}

#' Retrieve data. If data is paged, concatenate pages.
#' Only return the value without metadata.
#'
#' @param url Which url to fetch data from
#' @inheritDotParams retrieve_data
#' @export
#' @family retrieve
#'
#' @examples
#' \dontrun{
#' url <- "https://services.odata.org/V4/TripPinServiceRW/People"
#' retrieve_all(url)
#' }
retrieve_all <- function(url, ...) {
  pages <- list()
  next_link <- url

  i <- 1
  while (!is.null(next_link)) {
    data <- retrieve_data(next_link, ...)
    pages[[i]] <- data$value %||% data
    next_link <- data[["@odata.nextLink"]]
    i <- i + 1
  }

  if (is.data.frame(pages[[1]]))
    result <- do.call(rbind, pages)
  else
    result <- do.call(c, pages)

  result
}

#' Retrieve single instance.
#'
#' @param url Which url to fetch data from
#' @param default The default if nothing was found.
#' If not specified, an error is thrown in this case.
#' @inheritDotParams retrieve_data
#' @return Single value or default if none. If the result consists of
#' multiple records, an error is thrown.
#' @export
#' @family retrieve
#'
#' @examples
#' \dontrun{
#' url <- "https://services.odata.org/V4/TripPinServiceRW/People?$top=1"
#' retrieve_one(url)
#'
#' url <- "https://services.odata.org/V4/TripPinServiceRW/People('russellwhyte')"
#' retrieve_one(url)
#' }
retrieve_one <- function(url, default = stop("value not found"), ...) {
  data <- retrieve_data(url, ...)

  # Data is already a singleton
  if (is.null(data[["value"]]))
    return(data)

  # Make data a singleton
  value <- data$value
  n <- nrow(value) %||% length(value)

  if (n < 1)
    result <- default
  else if (n > 1)
    result <- stop("multiple values")
  else if (is.data.frame(value))
    result <- c(data[names(data) != "value"],
                lapply(value, function(col) col[[1]]))
  else
    result <- c(data[names(data) != "value"], value[[1]])

  result
}

#' Make an OData function available to R
#'
#' @description This turns an OData function into an R function
#' Parameters are serialized to json.
#' Scalar arguments should be passed as atomic vectors.
#' Array or object arguments should be passed as list.
#'
#' @param url Which url to fetch data from
#' @param metadata Whether and how metadata is included
#' @inheritDotParams retrieve_data
#' @return An R function
#' @export
#' @family retrieve
odata_function <- function(url, metadata = c("none", "minimal", "all"), ...) {
  force(metadata)
  outer_dots <- force(list(...))

  # Create a closure
  function(...) {
    args <- list(...)
    nargs <- names(args)

    left_hand <- ifelse(nchar(nargs) == 0, "", paste(nargs, "="))
    right_hand <- lapply(args, jsonlite::toJSON, auto_unbox = TRUE)
    arg_string <- paste(left_hand, right_hand, collapse = ",")
    encoded_args <- utils::URLencode(paste0("(", arg_string, ")"))

    url <- paste0(url, encoded_args)
    do.call(retrieve_data, c(url, outer_dots))
  }
}


R/utilities.R
#' Return x if not null, y otherwise (ruby style or)
#' @noRd
`%||%` <- function(x, y) {
  if (is.null(x))
    y
  else
    x
}

#' Create an odata_expr
#' @param x Character containing the odata code
#' @noRd
odata_expr <- function(x) {
  structure(x, class = "odata_expr")
}

#' Check if something is an odata expr
#' @param x The argument to check
#' @noRd
is.odata_expr <- function(x) {
  "odata_expr" %in% class(x)
}


#' Print an odata expr
#' @param expr Expression to print
#' @noRd
print.odata_expr <- function(expr) {
  cat("odata_expr: ", expr)
}

#' Concatenate two characters
#' @param s1 First character
#' @param s2 Second character
#' @noRd
`%s+%` <- function(s1, s2) {
  paste0(s1, s2)
}

#' Check whether something is a Date
#' @param x Argument to check
#' @noRd
is.Date <- function(x) {
  "Date" %in% class(x)
}


R/odata_query.R
# odata_query
# General methods to make OData url handling easier
#' @md


#' @title ODataQuery
#'
#' @description R6 class that represents an OData query
#'
#' @details
#'
#' This class has methods to build and navigate OData services:
#'
#' \itemize{
#' \item Use methods such as `$path()` and `$get()` to find a path.
#' \item Use methods such as `$select()` and `$filter()` to make your query.
#' \item Use methods such as `$retrieve()`, `$all()` and `$one()` to obtain
#' the results.
#' }
#' @export
ODataQuery <- R6::R6Class("ODataQuery",
  active = list(
    #' @field url Generate (encoded) url
    #' @param value Read-only
    url = function(value) {
      stopifnot(missing(value))
      query_string <- paste0(names(private$query_options), "=", private$query_options,
                             collapse = "&")

      if (length(private$query_options) > 0) {
        result <- paste0(private$service, private$resource, "?", query_string)
      } else {
        result <- paste0(private$service, private$resource)
      }
      utils::URLencode(result, repeated = TRUE)
    }
  ),

  public = list(
    #' @description Create a class representing a query.
    #'
    #' @param service The url of the endpoint to connect to.
    #' This url should not end with backslash.
    #' @param .resource Should not be used directly. Use $path() instead.
    #' @param .query_options Should not be used directly.
    #' Use methods such as $select(), $filter() and $query() instead.
    #' @param httr_args Additional parameters to pass to httr::GET
    #'
    #' @examples
    #' service <- ODataQuery$new("https://services.odata.org/V4/TripPinServiceRW")
    initialize = function(service, .resource = "", .query_options = list(), httr_args = list()) {
      stopifnot(length(service) == 1 && is.character(service))
      stopifnot(length(.resource) == 1 && is.character(.resource))
      stopifnot(is.list(.query_options))
      stopifnot(is.list(httr_args))

      private$service <- paste0(trimws(service, whitespace = "[/\\s]"), "/")
      private$resource <- .resource
      private$query_options <- .query_options
      private$httr_args <- httr_args
    },

    #' @description Print query, useful when debugging.
    #'
    #' @param top Number of records to print.
    #' @param ... Additional parameters are passed to print
    # inheritDotParams print
    #'
    #' @examples
    #' \dontrun{
    #' service <- ODataQuery$new("https://services.odata.org/V4/TripPinServiceRW")
    #' service$print(10)$path("People")$print()
    #' }
    print = function(top = 0, ...) {
      cat("ODataQuery:", self$url, "\n")

      if (top > 0) {
        df <- self$top(top)$retrieve(metadata = "minimal")
        print(df, ...)
      }

      invisible(self)
    },

    #' @description Supply path to the resource
    #'
    #' @param ... Components that lead to resource path
    #'
    #' @examples
    #' service <- ODataQuery$new("https://services.odata.org/V4/TripPinServiceRW")
    #' people_entity <- service$path("People")
    path = function(...) {
      resource <- paste(private$resource, ..., sep = "/")
      resource <- trimws(resource, whitespace = "[/\\s]")
      ODataQuery$new(private$service, resource, httr_args = private$httr_args)
    },

    #' @description Query an individual record by ID parameters
    #'
    #' @param ... ID-parameters (named or unnamed)
    #'
    #' @examples
    #' service <- ODataQuery$new("https://services.odata.org/V4/TripPinServiceRW")
    #' people_entity <- service$path("People")
    #' russellwhyte <- people_entity$get("russellwhyte")
    get = function(...) {
      args <- list(...)
      left_hand <- ifelse(nchar(names(args)) > 0,
                          paste0(names(args), "="),
                          names(args))
      right_hand <- lapply(args, represent_value)
      body <- paste0(left_hand, right_hand, collapse = ",")
      resource <- paste0(private$resource, "(", body, ")")
      ODataQuery$new(private$service, resource, httr_args = private$httr_args)
    },

    #' @description Path to an OData function
    #'
    #' @param fname Name of the function
    #' @param ... Options passed to retrieve_data
    # inheritDotParams retrieve_data
    #' @return closure
    #'
    #' @examples
    #' service <- ODataQuery$new("https://services.odata.org/V4/TripPinServiceRW")
    #' get_nearest_airport <- service$func('GetNearestAirport')
    #' \dontrun{
    #' get_nearest_airport(lat = 33, lon = -118)
    #' }
    func = function(fname, ...) {
      url <- paste(self$url, fname, sep = "/")
      odata_function(url, ..., httr_args = private$httr_args)
    },

    #' @description Supply custom query options that do not start with $
    #'
    #' @param ... Named lists where the names are custom query options
    #'
    #' @examples
    #' service <- ODataQuery$new("https://services.odata.org/V4/TripPinServiceRW")
    #' people_entity <- service$path("People")
    #' people_entity$query(filter = "FirstName eq 'scott'")$url
    query = function(...) {
      new_options <- list(...)
      query_options <- private$query_options
      query_options[names(new_options)] <- new_options
      return(ODataQuery$new(private$service, private$resource, query_options,
                            httr_args = private$httr_args))
    },

    #' @description Limit the number of results to n
    #'
    #' @param n Number of records to return at most
    #'
    #' @examples
    #' service <- ODataQuery$new("https://services.odata.org/V4/TripPinServiceRW")
    #' people_entity <- service$path("People")
    #' people_entity$top(10)
    top = function(n = 10) {
      stopifnot(is.numeric(n) && length(n) == 1)
      return(self$query(`$top` = n))
    },

    #' @description Skip first few items
    #'
    #' @param n Number of items to skip
    #'
    #' @examples
    #' service <- ODataQuery$new("https://services.odata.org/V4/TripPinServiceRW")
    #' people_entity <- service$path("People")
    #' people_entity$skip(10)
    skip = function(n = 10) {
      stopifnot(is.numeric(n) && length(n) == 1)
      return(self$query(`$skip` = n))
    },

    #' @description Select fields. If not present, all fields are returned.
    #'
    #' @param ... Fields to select
    #'
    #' @examples
    #' service <- ODataQuery$new("https://services.odata.org/V4/TripPinServiceRW")
    #' people_entity <- service$path("People")
    #' people_entity$select("FirstName", "LastName")
    select = function(...) {
      return(self$query(`$select` = paste(..., sep = ",")))
    },

    #' @description Apply filter to result
    #'
    #' @param ... Can be a raw odata query or query options. It's recommended to use
    #' query options because these will automatically escape parameters.
    #' The parameters are passed on to `and_query`.
    #' @inheritParams and_query()
    #' @seealso [and_query()] for details.
    #'
    #' @examples
    #' service <- ODataQuery$new("https://services.odata.org/V4/TripPinServiceRW")
    #' people_entity <- service$path("People")
    #' people_entity$filter(FirstName.eq = 'Scott')
    filter = function(...) {
      return(self$query(`$filter` = and_query(...)))
    },

    #' @description Expand on expansion properties
    #'
    #' @param ... Properties to extend on
    #'
    #' @examples
    #' service <- ODataQuery$new("https://services.odata.org/V4/TripPinServiceRW")
    #' people_entity <- service$path("People")
    #' people_entity$expand("Friends")
    expand = function(...) {
      return(self$query(`$expand` = paste(..., sep = ",")))
    },

    #' @description Order results by one or more keys
    #' @param ... Keys to order by. To order in descending order, the key can
    #' be prefixed by a negative sign.
    #'
    #' @examples
    #' service <- ODataQuery$new("https://services.odata.org/V4/TripPinServiceRW")
    #' people_entity <- service$path("People")
    #' people_entity$orderby('Concurrency')
    #' people_entity$orderby('-Concurrency')
    orderby = function(...) {
      keys <- c(...)
      orders <- ifelse(startsWith(keys, "-"),
                       paste(substr(keys, 2, 999), "desc"),
                       keys)
      orderby <- paste(orders, collapse = ",")
      return(self$query(`$orderby` = orderby))
    },

    #' @description Search the entity
    #'
    #' @param s Search string as defined by the endpoint.
    #'
    #' @examples
    #' service <- ODataQuery$new("https://services.odata.org/V4/TripPinServiceRW")
    #' people_entity <- service$path("People")
    #' people_entity$search('Boise')
    search = function(s) {
      return(self$query(`$search` = s))
    },

    #' @description Compute properties
    #'
    #' Add additional properties to query computed from other attributes.
    #'
    #' @param ... Named list of properties to compute
    #'
    #' @examples
    #' # Not really supported by this particular service.
    #' service <- ODataQuery$new("https://services.odata.org/V4/TripPinServiceRW")
    #' people_entity <- service$path("People")
    #' people_entity$compute(a = "5 MUL Concurrency")
    compute = function(...) {
      args <- list(...)
      right_hand <- ifelse(nchar(names(args)) == 0,
                           "",
                           paste("AS", names(args)))
      query <- paste(args, right_hand, collapse = ",")
      return(self$query(`$compute` = query))
    },

    #' @description Retrieve data
    #'
    #' @param count Whether to include a count of the total number of records
    #' @param ... Passed to retrieve_data
    #' @inheritParams retrieve_data(...)
    #'
    #' @examples
    #' \dontrun{
    #' service <- ODataQuery$new("https://services.odata.org/V4/TripPinServiceRW")
    #' people_entity$retrieve()
    #' }
    retrieve = function(count = FALSE, ...) {
      if (missing(count)) {
        url <- self$url
      } else {
        url <- self$query(`$count` = represent_value(count))$url
      }
      retrieve_data(url, ..., httr_args = private$httr_args)
    },

    #' @description Retrieve all data pages
    #'
    #' Return concatenation of value of all pages
    #'
    #' @param ... Passed to retrieve_all
    #' @inheritParams retrieve_all(...)
    #'
    #' @examples
    #' \dontrun{
    #' service <- ODataQuery$new("https://services.odata.org/V4/TripPinServiceRW")
    #' people_entity$all()
    #' people_entity$all(jsonlite_args = list(simplifyVector = False))
    #' }
    all = function(...) {
      retrieve_all(self$url, ..., httr_args = private$httr_args)
    },

    #' @description Retrieve individual
    #'
    #' @param ... Passed to retrieve_one
    #' @inheritParams retrieve_one(...)
    #'
    #' @examples
    #' \dontrun{
    #' service <- ODataQuery$new("https://services.odata.org/V4/TripPinServiceRW")
    #' people_entity$top(1)$one(default = NA)
    #' }
    one = function(...) {
      retrieve_one(self$url, ..., httr_args = private$httr_args)
    }
  ),

  private = list(
    # service Service endpoint
    service = NULL,

    # resource Resource name
    resource = NULL,

    # query_options Options to query on
    query_options = NULL,

    # httr_args Options to query on
    httr_args = NULL
  ),

  cloneable = FALSE
)


R/querying.R
# NSE-test

#' Macro to convert R to OData syntax
#'
#' @export
#' @param expr Expression to convert to OData
#'
#' @details
#' to_odata takes unquote R code and quotes its input. Use !! to unquote an argument.
#' to_odata_ requires its argument to be quoted already.
#'
#' Only a subset of R is supported.
#'
#' * arithmetic
#' The operators +, -, *, / and %% are mapped to add, sub, mul, divby and mod respectively
#'
#' * strings (characters in R)
#' toupper, tolower, startsWith, endsWith, nchar, paste, paste0, trimws
#'
#' * arrays (lists in R)
#' list, append, length
#'
#' * Formulae become lambdas in OData
#' (x ~ x$Name == "John")
#'
#' Every unknown function is passed as is.
#' If the function name is surrounded by percent signs it's treated as an infix operator.
#'
#' @examples
#' to_odata(Field == value)
#' address <- "Bakerstreet 4"
#' to_odata(!!address %in% Adresses)
#' to_odata(Friends$any(f ~ f$FirstName == 'John'))
#'
to_odata <- function(expr) {
  to_odata_(rlang::enexpr(expr))
}

#' @rdname to_odata
#' @export
to_odata_ <- function(expr) {
  expr <- process_dollar(expr)
  env <- odata_env(expr)
  represent_value(rlang::eval_bare(expr, env))
}

#' Create an environment to evaluate expression in
#' @param expr Expression for which to generate environment
#' @noRd
odata_env <- function(expr) {
  fnames <- all.names(expr, functions = TRUE)
  snames <- all.names(expr, functions = FALSE)
  fnames <- fnames[!(fnames %in% snames)]

  res <- new.env()

  for (fname in unique(fnames)) {
    if (startsWith(fname, "%") && endsWith(fname, "%"))
      res[[fname]] <- binop_function(" " %s+% substr(fname, 2, nchar(fname) - 1) %s+% " ")
    else
      res[[fname]] <- prefix_function(fname)
  }

  for (sname in unique(snames)) {
    res[[sname]] <- odata_expr(sname)
  }

  rlang::env_clone(standard_env, res)
}

#' Translate $ to /
#' @param expr Expression in which to replace dollar
#' @noRd
process_dollar <- function(expr) {
  if (rlang::is_call(expr)) {
    if (expr[[1]] == "$") {
      expr[[2]] <- process_dollar(expr[[2]])
      return (as.name(paste0(expr[[2]], "/", expr[[3]])))
    } else {
      expr[] <- lapply(expr[], process_dollar)
      return (expr)
    }
  } else {
    return (expr)
  }
}

#' Create a prefix function
#' @param f Function name in OData
#' @param ... Only used to suppress warning: ... may be used in an incorrect context
#' @noRd
prefix_function <- function(f, ...) {
  if (length(list(...) != 0))
    stop("prefix_function only takes one argument")

  rlang::new_function(
    rlang::exprs(... = ),
    rlang::expr(
      odata_expr(
        paste0(
          !!f, '(',
          paste(lapply(list(...), represent_value), collapse=','),
          ')')
        )
    ),
    rlang::caller_env()
  )
}

# Suppress warning: No visible binding for global variable e1, e2
e1 <- e2 <- NULL

#' Create an infix function
#' @param sep Separator to use in OData syntax
#' @noRd
binop_function <- function(sep) {
  rlang::new_function(
    rlang::exprs(e1 =, e2 =),
    rlang::expr(
      odata_expr(
        paste0(represent_value(e1), !!sep, represent_value(e2))
        )
    ),
    rlang::caller_env()
  )
}

#' Convert paste0 call to concat
#' @param ... Arguments to join
#' @noRd
odata_paste0 <- function(...) {
  Reduce(prefix_function('concat'), list(...))
}

#' Convert paste call to concat
#' @param ... Arguments to join
#' @param sep Separator to join arguments with
#' @noRd
odata_paste <- function(..., sep = " ") {
  Reduce(function(e1, e2) {odata_paste0(odata_paste0(e1, sep, e2))}, list(...))
}

#' Convert - call to odata. Can be used for subtraction or negation.
#' @param x First operand if y is passed, otherwise argument to negate.
#' @param y Second operand to minus function
#' @noRd
odata_minus <- function(x, y) {
  if (missing(y))
    prefix_function('-')(x)
  else
    binop_function(' sub ')(x, y)
}

#' Convert c call to OData
#'
#' Mainly used in combination with %in%
#' @param ... Contents of container
#' @noRd
odata_c <- function(...) {
  odata_expr(
    '(' %s+% paste(lapply(list(...), represent_value), collapse=",") %s+% ')'
  )
}

#' Convert list to an array
#' @param ... Contents of array
#' @noRd
odata_list <- function(...) {
  odata_expr(
    '[' %s+% paste(lapply(list(...), represent_value, double = TRUE), collapse=",") %s+% ']'
    )
}

#' Standard library of R functions and their OData equivalent
#' @noRd
standard_env <- rlang::child_env(
  .parent = rlang::empty_env(),
  `+` = binop_function(' add '),
  `-` = odata_minus,
  `*` = binop_function(' mul '),
  `/` = binop_function(' divby '),
  `%%` = binop_function(' mod '),
  `==` = binop_function(' eq '),
  `!=` = binop_function(' ne '),
  `<` = binop_function(' lt '),
  `>` = binop_function(' gt '),
  `<=` = binop_function(' le '),
  `>=` = binop_function(' ge '),
  `&&` = binop_function(' and '),
  `&` = binop_function(' and '),
  `||` = binop_function(' or '),
  `|` = binop_function(' or '),
  `!` = prefix_function('not'),
  paste0 = odata_paste0,
  paste = odata_paste,
  trimws = prefix_function('trim'),
  startsWith = prefix_function('startswith'),
  endsWith = prefix_function('endswith'),
  nchar = prefix_function("length"),
  `(` = prefix_function(""),
  `$` = binop_function('.'),
  `~` = binop_function(':'),
  list = odata_list,
  c = odata_c,
  `:` = function(e1, e2) {represent_value(as.list(e1:e2))},
  append = prefix_function(" concat "),
)

#' @title Create a combined filter
#' @export
#'
#' @param ... Raw odata queries or query options.
#'
#' @details
#' This function can be used with raw values or query options
#'
#' 1) Raw odata queries
#' Raw OData can be passed as string.
#' It's the responsibility of the caller that the argument is valid syntax
#' and values are escaped.
#'
#' 2) Query options
#' Query options can be passed as named parameters.
#'
#' Query options should be of the following form: `property.operator = value`
#'
#' * Property should be a property of the entity or individual.
#'
#' * Operation can have any of the following values:
#'
#'   * eq Whether property is equal to value.
#'   * ne Whether property is not equal to value.
#'   * gt Whether property is greater than value.
#'   * ge Whether property is greater than or equal to value.
#'   * lt Whether property is lower than value.
#'   * le Whether property is lower than or equal to value.
#'   * has Whether property has value as enumeration property.
#'   * startswith Whether property starts with value.
#'   * endswith Whether property ends with value.
#'   * contains Whether property contains value.
#'
#' * Value should be a string, double or boolean
#'   and will be escaped automatically.
#'
#'@md
#' @seealso <https://docs.oasis-open.org/odata/odata/v4.0/errata03/os/complete/part2-url-conventions/>
#'
#' @examples
#' and_query("Column eq OtherColumn",
#'           FirstName.startswith = 'A',
#'           LastName.eq = 'Scott')
#'
#' or_query("ExpireDate eq null",
#'          ExpireDate.lt = "2020-07-07")
#'
#' not_query(or_query(Age.lt = 21, Age.gt = 65))
#'
and_query <- function(...) {
  binop_query(" and ", ...)
}

#' @rdname and_query
#' @export
or_query <- function(...) {
  binop_query(" or ", ...)
}

#' @rdname and_query
#' @export
not_query <- function(...) {
  odata_expr(paste("not", and_query(...)))
}

#' Helper for creating and / or queries
#' @noRd
binop_query <- function(op, ...) {
  args <- list(...)

  # Find arg names
  if (is.null(names(args))) {
    argnames <- rep("", length(args))
  } else {
    argnames <- names(args)
  }

  query <- paste(Map(handle_parameter, argnames, args), collapse = op)
  odata_expr(paste0("(", query, ")"))
}


#' Handle parameter
#' @noRd
handle_parameter <- function(name, value) {
  INFIX <- c("eq", "ne", "gt", "ge", "lt", "le", "has")
  PREFIX <- c("startswith", "endswith", "contains")

  # Handle raw odata query
  if (is.null(name) || nchar(name) == 0)
    return(value)

  parts <- strsplit(name, ".", fixed = TRUE)[[1]]

  if (length(parts) != 2)
    stop("argument name must have format: \"property.operation\"")

  property <- parts[[1]]
  operator <- parts[[2]]

  if (operator %in% INFIX)
    result <- paste(property, operator, represent_value(value))
  else if (operator %in% PREFIX)
    result <- paste0(operator, "(", property, ",", represent_value(value), ")")
  else
    stop(paste0("Unknown operator \"", operator, "\". ",
                "Operator should be one of: ",
                paste0("\"", c(INFIX, PREFIX), "\"", collapse = ", ")), ".")

  result
}

#' Represent a primitive in OData (sanitized)
#' @param x Object to represent as odata query
#' @param double Whether a character should use double instead of single quotes
#' @noRd
represent_value <- function(x, double = FALSE) {
  if (is.odata_expr(x))
    return(x)
  else if (is.list(x))
    result <- jsonlite::toJSON(x, auto_unbox = TRUE)
  else if (length(x) > 1)
    result <- '(' %s+% paste(lapply(x, represent_value), collapse=",") %s+% ')'
  else if (is.name(x))
    result <- as.character(x)
  else if (is.Date(x))
    result <- as.character(x)
  else if (is.character(x)) {
    # Escape single or double quotes
    if (!double)
      result <- paste0("'", gsub("'", "''", x), "'")
    else
      result <- paste0("\"", gsub("\"", "\\\"", x), "\"")
  }
  else if (is.numeric(x))
    result <- x
  else if (is.logical(x))
    result <- tolower(x)
  else if (is.null(x))
    result <- "null"
  else {
    stop(paste("unknown type", class(x)))
  }

  odata_expr(result)
}
